src/compiler/crystal/syntax/lexer.cr

```diff
@@ -1780,61 +1780,103 @@ module Crystal
       next_char
 
       num = 0_u64
+      num_size = 0
       while true
         case next_char
         when '0'
-          num *= 2
+          num = num << 1
+          if num_size > 0
+            num_size += 1
+          end
         when '1'
-          num = num * 2 + 1
+          num = (num << 1) + 1
+          num_size += 1
         when '_'
           # Nothing
         else
           break
         end
       end
 
+      num = nil if num_size > 64
       finish_scan_prefixed_number num, negative, start
     end
 
     def scan_octal_number(start, negative)
       next_char
 
       num = 0_u64
-
+      num_size = 0
       while true
         char = next_char
         if '0' <= char <= '7'
-          num = num * 8 + (char - '0')
+          num = (num << 3) | (char - '0')
+          # First digit, if it's 2 through 9, adds a fake increment to the size.
+          if num_size == 0 && char > '1'
+            num_size += 1
+          end
+          if num_size > 0 || char != '0'
+            num_size += 1
+          end
         elsif char == '_'
         else
           break
         end
       end
 
+      # 0o177777_77777777_77777777 is the largest UInt64.
+      num = nil if num_size > 22 # or > 21 with first digit being 2 through 9
       finish_scan_prefixed_number num, negative, start
     end
 
     def scan_hex_number(start, negative = false)
       next_char
 
       num = 0_u64
+      num_size = 0
       while true
         char = next_char
         if char == '_'
         else
           hex_value = char_to_hex(char) { nil }
           if hex_value
-            num = num * 16 + hex_value
+            num = (num << 4) | hex_value
+            if num_size > 0 || char != '0'
+              num_size += 1
+            end
           else
             break
           end
         end
       end
 
+      # 0xFFFF_FFFF_FFFF_FFFF is the longest UInt64.
+      num = nil if num_size > 16
       finish_scan_prefixed_number num, negative, start
     end
 
-    def finish_scan_prefixed_number(num, negative, start)
+    def finish_scan_prefixed_number(num : Int?, negative : Bool, start : Int32)
+      if num.nil? # Doesn't even fit in UInt64
+        string_value = string_range_from_pool(start).gsub("_", "")
```

Comment:
Why modify the literal string with gsub?

Reply:
There's no other way to get rid of the underscores that might be there. I needed to get rid of trailing underscore, but figured that in the middle they're also not good. Do you suggest something else?

Reply:
I don't understand why you want to get rid of any underscores. I would expect to see the literal exactly as it's written.

Reply:
So, then notice this dichotomy:

```crystal
p 0x1234_1234_1234_i32
Error: 20014853001780 doesn't fit in an Int32
```
(that is the case both before and after^)

This is what you're suggesting:
```crystal
p 0x1234_1234_1234_1234_1234_i32
Error: 1234_1234_1234_1234_1234_ doesn't fit in an Int32
```
This is what I have for the moment
```crystal
p 0x1234_1234_1234_1234_1234_i32
Error: 12341234123412341234 doesn't fit in an Int32
```

Reply:
So, then notice this dichotomy:

```crystal
p 0x1234_1234_1234_i32
Error: 20014853001780 doesn't fit in an Int32
```
(that is the case both before and after^)

This is what you're suggesting:
```crystal
p 0x1234_1234_1234_1234_1234_i32
Error: 0x1234_1234_1234_1234_1234_ doesn't fit in an Int32
```
This is what I have for the moment
```crystal
p 0x1234_1234_1234_1234_1234_i32
Error: 0x12341234123412341234 doesn't fit in an Int32
```

Which do you prefer?

Reply:
We could also change even the pre-existing case to be like that.
Say,
```crystal
p 0x1234_1234_1234_i32
Error: 0x1234_1234_1234 doesn't fit in an Int32
```
But that's an even bigger change.

Reply:
I don't follow. The first error message is weird and confusing. It presents the literal in a different format than it is written. 

I would like to print the literal verbatim:
```cr
0x1234_1234_1234_1234_1234_i32 # => Error: 0x1234_1234_1234_1234_1234 doesn't fit in an Int32
```

Reply:
My main point is, I intended to not touch the case where the literal happens to fit UInt64. Because that's not what I originally opted to work on.
The example I show is pre-existing, doesn't change with the current state of this PR.

> So, then notice this dichotomy:
>
> ```crystal
> p 0x1234_1234_1234_i32
> Error: 20014853001780 doesn't fit in an Int32
> ```
> (that is the case both before and after^)


Do you think that should be changed too, as per
https://github.com/crystal-lang/crystal/pull/11447#discussion_r753382957 ?

Reply:
Yes, I would like that to change. It doesn't need to be in this PR, though.
But we should not go in a different direction with the changes that are in this PR.

Reply:
OK so fixed as per your comments, but not changing the pre-existing error message.

Reply:
Yeah, that's totally fine.